esp/services/ws_decoupledlogging/ws_decoupledloggingservice.cpp

```diff
@@ -89,64 +82,122 @@ void CWSDecoupledLogEx::init(IPropertyTree* cfg, const char* process, const char
             }
             loggingAgent->init(agentName, agentType, &loggingAgentTree, service);
             loggingAgent->initVariants(&loggingAgentTree);
-            IUpdateLogThread* logThread = createUpdateLogThread(&loggingAgentTree, service, agentName, tankFileDir, loggingAgent);
+            Owned<IUpdateLogThread> logThread = createUpdateLogThread(&loggingAgentTree, service, agentName, tankFileDir, loggingAgent);
             if(!logThread)
                 throw MakeStringException(-1, "Failed to create update log thread for %s", agentName);
-    
+
             CLogRequestReader* logRequestReader = logThread->getLogRequestReader();
             if (!logRequestReader)
                 throw MakeStringException(-1, "CLogRequestReader not found for %s.", agentName);
-            group->loggingAgentThreads.push_back(logThread);
+
+            if (group->getLoggingAgentThread(agentName))
+                throw MakeStringException(-1, "%s: >1 logging agents are named as %s.", groupName, agentName);
+            group->addLoggingAgentThread(agentName, logThread);
         }
-        logGroups.push_back(group);
+        logGroups.insert({groupName, group});
     }
 }
 
 bool CWSDecoupledLogEx::onGetLogAgentSetting(IEspContext& context, IEspGetLogAgentSettingRequest& req, IEspGetLogAgentSettingResponse& resp)
 {
     try
     {
-        BoolHash logAgentFound;
-        StringBuffer errorStatus;
-        StringArray& agentNames = req.getAgentNames();
-        IArrayOf<IEspLogAgentSetting> logAgentSettings;
-        for (auto gp : logGroups)
+        IArrayOf<IEspLogAgentGroupSetting> groupSettingResp;
+        IArrayOf<IConstLogAgentGroup>& groups = req.getGroups();
+        if (!groups.ordinality())
```

Comment:
This seems more difficult to follow than needed. I figured out what it's doing, but think it could be structured differently for readability. Specifically:
- There are two ways to select a group. Once the group is identified, pass the group and a StringArray pointer representing the selected agents (or nullptr when ordinality is one) to a separate method.
- When processing a group, there are two ways to select an agent. Pass a selected agent and any other required information into a separate method.

Making sure two independent implementations stay in sync could be error prone.